# Employment data Processing #


## Job Data Analysis for MRA Region

This notebook processes two datasets regarding Employment in the MRA region:

A. **Total Number of Jobs in a PC4 Zone**

B. **Number of Jobs across each sector in a PC4 Zone**
     

# **A. Total Number of Jobs** 

## **1. Import Libraries and Loading Data**

## 1.1 Importing Libraries

In [108]:
import pandas as pd
import numpy as np

## A1.2 Loading Data

**Source:** Data is obtained upon request by emailing the Onderzoek en Statistiek (Research and Statistics) team of Amsterdam at redactie.os@amsterdam.nl.

**Contact Person at the time:** Rogier Van der Groep (rogier.van.der.groep@amsterdam.nl)

In [109]:
MRA_total_employment_file =  'lisa_18_23_pc4_jobs Total.xlsx'
MRA_total_employment =  pd.read_excel(MRA_total_employment_file)



In [110]:
MRA_total_employment

peildatum   pc4  gemeentecode gemeentenaam  banen
0    2018-04-01  1000           363    Amsterdam      1
1    2018-04-01  1011           363    Amsterdam  13884
2    2018-04-01  1012           363    Amsterdam  31651
3    2018-04-01  1013           363    Amsterdam  19035
4    2018-04-01  1014           363    Amsterdam  11803
...         ...   ...           ...          ...    ...
2195 2023-04-01  8241           995     Lelystad    235
2196 2023-04-01  8242           995     Lelystad   2452
2197 2023-04-01  8243           995     Lelystad   5432
2198 2023-04-01  8244           995     Lelystad    669
2199 2023-04-01  8245           995     Lelystad   1204

[2200 rows x 5 columns]

In [111]:
MRA_total_employment['peildatum'].dtype

dtype('<M8[ns]')

### Job Registration and ODiN Survey Timing

- Jobs are registered annually on April 1st. For example, jobs registered on April 1, 2018, reflect the data for the year 2018.
- The ODiN survey, however, can occur at any time during the year. Therefore, to align with the most recent job data, it is preferable to use the job data from April 1st of the following year.

For instance:
- For ODiN data collected in 2018, use the jobs data from April 1, 2019.
- Similarly, for ODiN data collected in 2022, use the jobs data from April 1, 2023.


In [112]:

MRA_total_employment_2018 =  MRA_total_employment[MRA_total_employment['peildatum'].dt.year == 2019]
MRA_total_employment_2022 =  MRA_total_employment[MRA_total_employment['peildatum'].dt.year == 2023]

In [113]:
MRA_total_employment_2018

peildatum   pc4  gemeentecode gemeentenaam  banen
366 2019-04-01  1000           363    Amsterdam      1
367 2019-04-01  1011           363    Amsterdam  13550
368 2019-04-01  1012           363    Amsterdam  32506
369 2019-04-01  1013           363    Amsterdam  20589
370 2019-04-01  1014           363    Amsterdam  13732
..         ...   ...           ...          ...    ...
726 2019-04-01  8241           995     Lelystad    208
727 2019-04-01  8242           995     Lelystad   2388
728 2019-04-01  8243           995     Lelystad   5114
729 2019-04-01  8244           995     Lelystad    554
730 2019-04-01  8245           995     Lelystad    802

[365 rows x 5 columns]

In [114]:
MRA_total_employment_2022

peildatum   pc4  gemeentecode gemeentenaam  banen
1827 2023-04-01  1000           363    Amsterdam      2
1828 2023-04-01  1011           363    Amsterdam  14203
1829 2023-04-01  1012           363    Amsterdam  31692
1830 2023-04-01  1013           363    Amsterdam  23203
1831 2023-04-01  1014           363    Amsterdam  14759
...         ...   ...           ...          ...    ...
2195 2023-04-01  8241           995     Lelystad    235
2196 2023-04-01  8242           995     Lelystad   2452
2197 2023-04-01  8243           995     Lelystad   5432
2198 2023-04-01  8244           995     Lelystad    669
2199 2023-04-01  8245           995     Lelystad   1204

[373 rows x 5 columns]

## **2. Data Processing**

In [115]:
# Grouping the MRA_total_employment_2018 DataFrame by 'pc4' to aggregate job data
MRA_total_employment_pc4_2018 = MRA_total_employment_2018.groupby('pc4').agg({
    # Sum the total number of jobs ('banen') for each 'pc4' zone
    'banen': 'sum',
    # Take the first occurrence of 'gemeentecode' for each 'pc4' zone
    'gemeentecode': 'first',
    # Take the first occurrence of 'gemeentenaam' for each 'pc4' zone
    'gemeentenaam': 'first'
}).reset_index()  # Reset the index to turn the grouped data back into a DataFrame
MRA_total_employment_pc4_2018

pc4  banen  gemeentecode gemeentenaam
0    1000      1           363    Amsterdam
1    1011  13550           363    Amsterdam
2    1012  32506           363    Amsterdam
3    1013  20589           363    Amsterdam
4    1014  13732           363    Amsterdam
..    ...    ...           ...          ...
359  8241    208           995     Lelystad
360  8242   2388           995     Lelystad
361  8243   5114           995     Lelystad
362  8244    554           995     Lelystad
363  8245    802           995     Lelystad

[364 rows x 4 columns]

In [116]:
# Similarly grouping by Gemeentee (Municipality)

MRA_total_employment_gem_2018 = MRA_total_employment_2018.groupby('gemeentecode').agg({
    'banen': 'sum',
    'gemeentenaam': 'first'
}).reset_index()
MRA_total_employment_gem_2018

gemeentecode   banen    gemeentenaam
0             34   85171          Almere
1            358   16069        Aalsmeer
2            362   52505      Amstelveen
3            363  699964       Amsterdam
4            375   18244       Beverwijk
5            376    3795        Blaricum
6            377    6385     Bloemendaal
7            384   19308          Diemen
8            385   15609   Edam-Volendam
9            392   70255         Haarlem
10           394  158933  Haarlemmermeer
11           396    9628       Heemskerk
12           397    8736       Heemstede
13           402   48195       Hilversum
14           406   12933          Huizen
15           415    3673       Landsmeer
16           417    4739           Laren
17           431    4048        Oostzaan
18           437   15484    Ouder-Amstel
19           439   31175       Purmerend
20           450    3968        Uitgeest
21           451   13002        Uithoorn
22           453   34171          Velsen
23           473    5601       Zandvoort
24           479   66539        Zaanstad
25           852    4629       Waterland
26           880    5086      Wormerland
27           995   36546        Lelystad
28          1696    8845      Wijdemeren
29          1942   21490    Gooise Meren

In [117]:
# Doing the same for 2022 

# PC4 grouping 
MRA_total_employment_pc4_2022 = MRA_total_employment_2022.groupby('pc4').agg({
    'banen': 'sum',
    'gemeentecode': 'first',
    'gemeentenaam': 'first'
}).reset_index()
MRA_total_employment_pc4_2022

pc4  banen  gemeentecode gemeentenaam
0    1000      2           363    Amsterdam
1    1011  14203           363    Amsterdam
2    1012  31692           363    Amsterdam
3    1013  23203           363    Amsterdam
4    1014  14759           363    Amsterdam
..    ...    ...           ...          ...
366  8241    235           995     Lelystad
367  8242   2452           995     Lelystad
368  8243   5432           995     Lelystad
369  8244    669           995     Lelystad
370  8245   1204           995     Lelystad

[371 rows x 4 columns]

In [118]:
# Gemeente Grouping

MRA_total_employment_gem_2022 = MRA_total_employment_2022.groupby('gemeentecode').agg({
    'banen': 'sum',
    'gemeentenaam': 'first'
}).reset_index()
MRA_total_employment_gem_2022

gemeentecode   banen    gemeentenaam
0             34   97738          Almere
1            358   17390        Aalsmeer
2            362   55655      Amstelveen
3            363  746027       Amsterdam
4            375   18973       Beverwijk
5            376    4512        Blaricum
6            377    6838     Bloemendaal
7            384   19269          Diemen
8            385   16179   Edam-Volendam
9            392   77554         Haarlem
10           394  168129  Haarlemmermeer
11           396   10532       Heemskerk
12           397    9664       Heemstede
13           402   52038       Hilversum
14           406   13773          Huizen
15           415    4018       Landsmeer
16           417    4890           Laren
17           431    4363        Oostzaan
18           437   17469    Ouder-Amstel
19           439   34551       Purmerend
20           450    4124        Uitgeest
21           451   13928        Uithoorn
22           453   37723          Velsen
23           473    6145       Zandvoort
24           479   72542        Zaanstad
25           852    4951       Waterland
26           880    5329      Wormerland
27           995   40365        Lelystad
28          1696    9251      Wijdemeren
29          1942   24483    Gooise Meren

### PC4 Zones Comparison Between 2018 and 2022

- In 2022, there are 371 PC4 zones, while in 2018 there were 364 PC4 zones within the MRA region.
- Therefore, it is necessary to identify which PC4 zones are present in 2022 but were not included in 2018.







In [119]:
 #Convert pc4 column to sets for both dataframes
pc4_2022_set = set(MRA_total_employment_pc4_2022['pc4'])
pc4_2018_set = set(MRA_total_employment_pc4_2018['pc4'])

# Find values in pc4_2022_set that are not in pc4_2018_set
values_not_in_2018 = pc4_2022_set - pc4_2018_set

print("Values in MRA_total_employment_pc4_2022 not present in MRA_total_employment_pc4_2018:")
print(values_not_in_2018)

Values in MRA_total_employment_pc4_2022 not present in MRA_total_employment_pc4_2018:
{1320, 2130, 1300, 1301, 1302, 1305, 1084}


### Handling PC4 Zones Present Only in 2023

- Some PC4 zones that appear in the 2023 data have very low employment numbers and do not exist in the 2018 dataset.
- These PC4 zones include:

    | Peildatum    | PC4  | Gemeentecode | Gemeentenaam   | Banen |
    |--------------|------|--------------|----------------|-------|
    | 04/01/2023   | 1084 | 0363         | Amsterdam      | 1     |
    | 04/01/2023   | 1300 | 0034         | Almere         | 1     |
    | 04/01/2023   | 1301 | 0034         | Almere         | 1     |
    | 04/01/2023   | 1305 | 0034         | Almere         | 5     |
    | 04/01/2023   | 1320 | 0034         | Almere         | 1     |
    | 04/01/2023   | 2130 | 0394         | Haarlemmermeer | 1     |

- Since these PC4 zones are new and have very low employment numbers, they can be removed.



In [120]:
# Identify the PC4 zones that are common between the 2022 and 2018 datasets
common_pc4 = MRA_total_employment_pc4_2022['pc4'].isin(MRA_total_employment_pc4_2018['pc4'])

# Filter the 2022 dataset to retain only those PC4 zones that are also present in the 2018 dataset
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022[common_pc4]
MRA_total_employment_pc4_2022

pc4  banen  gemeentecode gemeentenaam
0    1000      2           363    Amsterdam
1    1011  14203           363    Amsterdam
2    1012  31692           363    Amsterdam
3    1013  23203           363    Amsterdam
4    1014  14759           363    Amsterdam
..    ...    ...           ...          ...
366  8241    235           995     Lelystad
367  8242   2452           995     Lelystad
368  8243   5432           995     Lelystad
369  8244    669           995     Lelystad
370  8245   1204           995     Lelystad

[364 rows x 4 columns]

In [121]:
# Rename the 'banen' column to 'Total Employment 2018-19' in the MRA_total_employment_pc4_2018 DataFrame

MRA_total_employment_pc4_2018.rename(columns={'banen': 'Total Employment 2018-19'})

pc4  Total Employment 2018-19  gemeentecode gemeentenaam
0    1000                         1           363    Amsterdam
1    1011                     13550           363    Amsterdam
2    1012                     32506           363    Amsterdam
3    1013                     20589           363    Amsterdam
4    1014                     13732           363    Amsterdam
..    ...                       ...           ...          ...
359  8241                       208           995     Lelystad
360  8242                      2388           995     Lelystad
361  8243                      5114           995     Lelystad
362  8244                       554           995     Lelystad
363  8245                       802           995     Lelystad

[364 rows x 4 columns]

In [122]:
#Similarly for 2022 data

MRA_total_employment_pc4_2022.rename(columns={'banen': 'Total Employment 2022-23'})

pc4  Total Employment 2022-23  gemeentecode gemeentenaam
0    1000                         2           363    Amsterdam
1    1011                     14203           363    Amsterdam
2    1012                     31692           363    Amsterdam
3    1013                     23203           363    Amsterdam
4    1014                     14759           363    Amsterdam
..    ...                       ...           ...          ...
366  8241                       235           995     Lelystad
367  8242                      2452           995     Lelystad
368  8243                      5432           995     Lelystad
369  8244                       669           995     Lelystad
370  8245                      1204           995     Lelystad

[364 rows x 4 columns]

In [123]:
# Now ensure the columns are of the same dtype!
# Just to be on the safe side, convert gem_code and pc4 code to object type.
# to ensure that they don't sum up at all while aggregation by mistake and if you want to 
# expand them into pc6 codes


# Step 1: Convert column to object type
#2018
MRA_total_employment_pc4_2018['pc4'] = MRA_total_employment_pc4_2018['pc4'].astype('int64')
MRA_total_employment_pc4_2018['gemeentecode'] = MRA_total_employment_pc4_2018['gemeentecode'].astype('int64')
#2022
MRA_total_employment_pc4_2022['pc4'] = MRA_total_employment_pc4_2022['pc4'].astype('int64')
MRA_total_employment_pc4_2022['gemeentecode'] = MRA_total_employment_pc4_2022['gemeentecode'].astype('int64')




# Step 2: Remove NaN values from 'pc4' column
#2018
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018.dropna(subset=['pc4'])
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018.dropna(subset = ['gemeentecode'])
#2022
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022.dropna(subset=['pc4'])
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022.dropna(subset = ['gemeentecode'])

# Step 3: Remove rows where 'pc4' is 0
#2018
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018[MRA_total_employment_pc4_2018['pc4'] != 0]
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018[MRA_total_employment_pc4_2018['gemeentecode'] != 0]
#2022
RA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022[MRA_total_employment_pc4_2022['pc4'] != 0]
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022[MRA_total_employment_pc4_2022['gemeentecode'] != 0]



# Verify the changes
#2018
print(MRA_total_employment_pc4_2018['pc4'].dtype)  # Should show 'int64'
print(MRA_total_employment_pc4_2018['pc4'].isnull().sum())  # Should show 0
print((MRA_total_employment_pc4_2018['pc4'] == 0).sum())  # Should show 0
print('Gemeente Code')
print(MRA_total_employment_pc4_2018['gemeentecode'].dtype)
print(MRA_total_employment_pc4_2018['gemeentecode'].isnull().sum())
print((MRA_total_employment_pc4_2018['gemeentecode'] == 0).sum())

#2022
print(MRA_total_employment_pc4_2022['pc4'].dtype)  # Should show 'object'
print(MRA_total_employment_pc4_2022['pc4'].isnull().sum())  # Should show 0
print((MRA_total_employment_pc4_2022['pc4'] == 0).sum())  # Should show 0
print('Gemeente Code')
print(MRA_total_employment_pc4_2022['gemeentecode'].dtype)
print(MRA_total_employment_pc4_2022['gemeentecode'].isnull().sum())
print((MRA_total_employment_pc4_2022['gemeentecode'] == 0).sum())



int64
0
0
Gemeente Code
int64
0
0
int64
0
0
Gemeente Code
int64
0
0


C:\Users\rohan\AppData\Local\Temp\ipykernel_37356\3333697854.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  MRA_total_employment_pc4_2022['pc4'] = MRA_total_employment_pc4_2022['pc4'].astype('int64')
C:\Users\rohan\AppData\Local\Temp\ipykernel_37356\3333697854.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  MRA_total_employment_pc4_2022['gemeentecode'] = MRA_total_employment_pc4_2022['gemeentecode'].astype('int64')


In [124]:
# Renaming retaining only relevant columns for clarity 

#2018
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018.rename(columns={'pc4': 'PC4 Code', 'banen': 'Total Employment 2018-19', 'gemeentecode' : 'Gemeente Code', 'gemeentenaam': 'Gemeente Name' })
MRA_total_employment_pc4_2018 = MRA_total_employment_pc4_2018[['PC4 Code', 'Gemeente Code', 'Gemeente Name', 'Total Employment 2018-19'] ]

#2022
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022.rename(columns={'pc4': 'PC4 Code', 'banen': 'Total Employment 2022-23', 'gemeentecode' : 'Gemeente Code', 'gemeentenaam': 'Gemeente Name' })
MRA_total_employment_pc4_2022 = MRA_total_employment_pc4_2022[['PC4 Code', 'Gemeente Code', 'Gemeente Name', 'Total Employment 2022-23'] ]

MRA_total_employment_pc4_2018

PC4 Code  Gemeente Code Gemeente Name  Total Employment 2018-19
0        1000            363     Amsterdam                         1
1        1011            363     Amsterdam                     13550
2        1012            363     Amsterdam                     32506
3        1013            363     Amsterdam                     20589
4        1014            363     Amsterdam                     13732
..        ...            ...           ...                       ...
359      8241            995      Lelystad                       208
360      8242            995      Lelystad                      2388
361      8243            995      Lelystad                      5114
362      8244            995      Lelystad                       554
363      8245            995      Lelystad                       802

[364 rows x 4 columns]

# **B. Number of Jobs across each sector** 

## B1.2 Loading Data


In [125]:
MRA_sector_employment_file = 'lisa_18_23_pc4_jobs Per sector.xlsx'
MRA_sector_employment = pd.read_excel(MRA_sector_employment_file)
MRA_sector_employment

peildatum   pc4  gemeentecode gemeentenaam sbi_sectie_code  \
0     2018-04-01  1000           363    Amsterdam               J   
1     2018-04-01  1011           363    Amsterdam               A   
2     2018-04-01  1011           363    Amsterdam               B   
3     2018-04-01  1011           363    Amsterdam               C   
4     2018-04-01  1011           363    Amsterdam               D   
...          ...   ...           ...          ...             ...   
35913 2023-04-01  8245           995     Lelystad               P   
35914 2023-04-01  8245           995     Lelystad               Q   
35915 2023-04-01  8245           995     Lelystad               R   
35916 2023-04-01  8245           995     Lelystad               S   
35917 2023-04-01  8245           995     Lelystad             NaN   

                                 sbi_sectie_omschrijving  banen  
0                             Informatie en communicatie      0  
1                          Landbouw, bosbouw en visserij      0  
2                               Winning van delfstoffen       0  
3                                              Industrie     70  
4      Productie en distributie van en handel in elek...      0  
...                                                  ...    ...  
35913                                          Onderwijs    130  
35914                       Gezondheids- en welzijnszorg    160  
35915                        Cultuur, sport en recreatie    120  
35916                            Overige dienstverlening     40  
35917                                                NaN      0  

[35918 rows x 7 columns]

## Processing Data 

In [126]:
# Creating Separate database for 2018 and 2022
MRA_sector_employment_2018 = MRA_sector_employment[MRA_sector_employment['peildatum'].dt.year == 2019]


MRA_sector_employment_2022 = MRA_sector_employment[MRA_sector_employment['peildatum'].dt.year == 2023]
MRA_sector_employment_2018.columns


Index(['peildatum', 'pc4', 'gemeentecode', 'gemeentenaam', 'sbi_sectie_code',
       'sbi_sectie_omschrijving', 'banen'],
      dtype='object')

In [127]:
MRA_sector_employment_2018

peildatum   pc4  gemeentecode gemeentenaam sbi_sectie_code  \
5918  2019-04-01  1000           363    Amsterdam               J   
5919  2019-04-01  1011           363    Amsterdam               A   
5920  2019-04-01  1011           363    Amsterdam               B   
5921  2019-04-01  1011           363    Amsterdam               C   
5922  2019-04-01  1011           363    Amsterdam               D   
...          ...   ...           ...          ...             ...   
11849 2019-04-01  8245           995     Lelystad               P   
11850 2019-04-01  8245           995     Lelystad               Q   
11851 2019-04-01  8245           995     Lelystad               R   
11852 2019-04-01  8245           995     Lelystad               S   
11853 2019-04-01  8245           995     Lelystad             NaN   

                                 sbi_sectie_omschrijving  banen  
5918                          Informatie en communicatie      0  
5919                       Landbouw, bosbouw en visserij      0  
5920                            Winning van delfstoffen       0  
5921                                           Industrie     70  
5922   Productie en distributie van en handel in elek...      0  
...                                                  ...    ...  
11849                                          Onderwijs     90  
11850                       Gezondheids- en welzijnszorg    160  
11851                        Cultuur, sport en recreatie     60  
11852                            Overige dienstverlening     30  
11853                                                NaN      0  

[5936 rows x 7 columns]

In [128]:
# Pivoting, so that for each pc4, the sum of jobs per sector is summed up for 2018 and 2022

#2018
pivoted_MRA_sector_employment_2018 = MRA_sector_employment_2018.pivot_table(
    index=["peildatum", "pc4", "gemeentecode", "gemeentenaam"], 
    columns="sbi_sectie_omschrijving", 
    values="banen", 
    aggfunc="sum"
).reset_index()


#2022

pivoted_MRA_sector_employment_2022 = MRA_sector_employment_2022.pivot_table(
    index=["peildatum", "pc4", "gemeentecode", "gemeentenaam"], 
    columns="sbi_sectie_omschrijving", 
    values="banen", 
    aggfunc="sum"
).reset_index()

pivoted_MRA_sector_employment_2018



sbi_sectie_omschrijving  peildatum   pc4  gemeentecode gemeentenaam  \
0                       2019-04-01  1000           363    Amsterdam   
1                       2019-04-01  1011           363    Amsterdam   
2                       2019-04-01  1012           363    Amsterdam   
3                       2019-04-01  1013           363    Amsterdam   
4                       2019-04-01  1014           363    Amsterdam   
..                             ...   ...           ...          ...   
360                     2019-04-01  8241           995     Lelystad   
361                     2019-04-01  8242           995     Lelystad   
362                     2019-04-01  8243           995     Lelystad   
363                     2019-04-01  8244           995     Lelystad   
364                     2019-04-01  8245           995     Lelystad   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                      NaN                            
1                                                                   1870.0                            
2                                                                   2330.0                            
3                                                                   5070.0                            
4                                                                   2180.0                            
..                                                                     ...                            
360                                                                   80.0                            
361                                                                  160.0                            
362                                                                  290.0                            
363                                                                  120.0                            
364                                                                  130.0                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                   NaN                          NaN   
1                                  80.0                       1880.0   
2                                  60.0                       1420.0   
3                                 600.0                       1250.0   
4                                 770.0                        230.0   
..                                  ...                          ...   
360                                10.0                         20.0   
361                                70.0                        100.0   
362                               490.0                         60.0   
363                                50.0                         40.0   
364                                60.0                         60.0   

sbi_sectie_omschrijving  Detailhandel  \
0                                 NaN   
1                               930.0   
2                              7820.0   
3                              1550.0   
4                               460.0   
..                                ...   
360                              10.0   
361                            1120.0   
362                             560.0   
363                              40.0   
364                             120.0   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                               NaN   
1                                                               NaN   
2                                                               NaN   
3                                                               NaN   
4                                                               0.0   
..                                                              ...   
360                                                    

In [129]:
# Filling Nan values as 0 and converting the sector job columns as int

pivoted_MRA_sector_employment_2018 = pivoted_MRA_sector_employment_2018.fillna(0).astype({col: np.int64 for col in pivoted_MRA_sector_employment_2018.columns if col not in ["peildatum", "pc4", "gemeentecode", "gemeentenaam"]})
pivoted_MRA_sector_employment_2022 = pivoted_MRA_sector_employment_2022.fillna(0).astype({col: np.int64 for col in pivoted_MRA_sector_employment_2022 if col not in         ["peildatum", "pc4", "gemeentecode", "gemeentenaam"]})

pivoted_MRA_sector_employment_2022

sbi_sectie_omschrijving  peildatum   pc4  gemeentecode gemeentenaam  \
0                       2023-04-01  1000           363    Amsterdam   
1                       2023-04-01  1011           363    Amsterdam   
2                       2023-04-01  1012           363    Amsterdam   
3                       2023-04-01  1013           363    Amsterdam   
4                       2023-04-01  1014           363    Amsterdam   
..                             ...   ...           ...          ...   
368                     2023-04-01  8241           995     Lelystad   
369                     2023-04-01  8242           995     Lelystad   
370                     2023-04-01  8243           995     Lelystad   
371                     2023-04-01  8244           995     Lelystad   
372                     2023-04-01  8245           995     Lelystad   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                        0                            
1                                                                     2030                            
2                                                                     2290                            
3                                                                     6060                            
4                                                                     2750                            
..                                                                     ...                            
368                                                                     90                            
369                                                                    210                            
370                                                                    340                            
371                                                                    130                            
372                                                                    180                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                     0                            0   
1                                    90                         1830   
2                                    70                         1530   
3                                   560                         1320   
4                                   420                          330   
..                                  ...                          ...   
368                                  10                            0   
369                                 110                           80   
370                                 460                           60   
371                                  80                           30   
372                                 140                          120   

sbi_sectie_omschrijving  Detailhandel  \
0                                   0   
1                                 940   
2                                6770   
3                                1620   
4                                 500   
..                                ...   
368                                20   
369                              1010   
370                               700   
371                                60   
372                               210   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                                 0   
1                                                                 0   
2                                                                 0   
3                                                                 0   
4                                                                 0   
..                                                              ...   
368                                                    

In [130]:
pivoted_MRA_sector_employment_2022.columns

Index(['peildatum', 'pc4', 'gemeentecode', 'gemeentenaam',
       'Advisering, onderzoek en overige specialistische zakelijke dienstverlening',
       'Bouwnijverheid', 'Cultuur, sport en recreatie', 'Detailhandel',
       'Extraterritoriale organisaties en lichamen', 'Financiële instellingen',
       'Gezondheids- en welzijnszorg', 'Groothandel', 'Industrie',
       'Informatie en communicatie', 'Landbouw, bosbouw en visserij',
       'Logies-, maaltijd- en drankverstrekking', 'Onderwijs',
       'Openbaar bestuur, overheidsdiensten en verplichte sociale verzekeringen',
       'Overige dienstverlening',
       'Productie en distributie van en handel in elektriciteit, aardgas, stoom en gekoelde lucht',
       'Verhuur van en handel in onroerend goed',
       'Verhuur van roerende goederen en overige zakelijke dienstverlening',
       'Vervoer en opslag',
       'Winning en distributie van water: afval- en afvalwaterbeheer en sanering',
       'Winning van delfstoffen '],
      dtype='o

In [131]:

# List of columns to exclude from the sum
exclude_columns = ['peildatum', 'pc4', 'gemeentecode', 'gemeentenaam', 'TOTAL', 'Total Across Categories'] #Total and 'Total Across Categories' are added 
                                                                                                           #to ensure it doesn't sum again while merging with Masterdata

# Create a list of columns to sum by excluding the above columns
columns_to_sum_18 = [col for col in pivoted_MRA_sector_employment_2018.columns if col not in exclude_columns]
columns_to_sum_22 = [col for col in pivoted_MRA_sector_employment_2022.columns if col not in exclude_columns]


# Sum the values of the specified columns and add a new column 'TOTAL' to the DataFrame
pivoted_MRA_sector_employment_2018['TOTAL'] = pivoted_MRA_sector_employment_2018[columns_to_sum_18].sum(axis=1)
pivoted_MRA_sector_employment_2022['TOTAL'] = pivoted_MRA_sector_employment_2022[columns_to_sum_22].sum(axis=1)


# Display the updated DataFrame
#print(pivoted_MRA_sector_employment_2018)


#NOTE :Blank values are converted to  NaN values are excluded, therefore are excluded.
# The numbers here won't exactly match the total pc4 values in MRA_total_employment_file 
# The reason is, due to GDPR laws, very small values are put as zeroes across sector jobs




#Checking if the the table is pivoted correctly by checking the total column with the sum of the excel file for any random postcode


pivoted_MRA_sector_employment_2018[pivoted_MRA_sector_employment_2018['pc4'] == 2135]
# In excel, lisa_18_23_pc4_jobs Per sector filter the peildatum to 2019 and pc4 to 2135 and sum up the banen, it should sum to the value of the last 'TOTAL' column of the dataframe

sbi_sectie_omschrijving  peildatum   pc4  gemeentecode    gemeentenaam  \
330                     2019-04-01  2135           394  Haarlemmermeer   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
330                                                                    260                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
330                                 100                           50   

sbi_sectie_omschrijving  Detailhandel  \
330                               470   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
330                                                               0   

sbi_sectie_omschrijving  Financiële instellingen  ...  Onderwijs  \
330                                           10  ...        400   

sbi_sectie_omschrijving  Openbaar bestuur, overheidsdiensten en verplichte sociale verzekeringen  \
330                                                                      0                         

sbi_sectie_omschrijving  Overige dienstverlening  \
330                                           70   

sbi_sectie_omschrijving  Productie en distributie van en handel in elektriciteit, aardgas, stoom en gekoelde lucht  \
330                                                                      0                                           

sbi_sectie_omschrijving  Verhuur van en handel in onroerend goed  \
330                                                           10   

sbi_sectie_omschrijving  Verhuur van roerende goederen en overige zakelijke dienstverlening  \
330                                                                    140                    

sbi_sectie_omschrijving  Vervoer en opslag  \
330                                     50   

sbi_sectie_omschrijving  Winning en distributie van water: afval- en afvalwaterbeheer en sanering  \
330                                                                      0                          

sbi_sectie_omschrijving  Winning van delfstoffen   TOTAL  
330                                             0   2040  

[1 rows x 26 columns]

In [132]:
#Ensuring that only one gementee code and gemeentenaam is there for each pc4 

# 2018
# Create a dictionary for aggregation
agg_dict = {col: 'sum' for col in columns_to_sum_18}
agg_dict['gemeentecode'] = 'first'  # Adding gemeentecode to keep the first entry
agg_dict['gemeentenaam'] = 'first'  # Adding gemeentenaam to keep the first entry
agg_dict['TOTAL'] = 'first'

# Perform the group by and aggregate functions
pivoted_MRA_sector_employment_pc4_2018 = pivoted_MRA_sector_employment_2018.groupby('pc4').agg(agg_dict).reset_index()


# 2022
agg_dict_22 = {col: 'sum' for col in columns_to_sum_22}
agg_dict_22['gemeentecode'] = 'first'  # Adding gemeentecode to keep the first entry
agg_dict_22['gemeentenaam'] = 'first'  # Adding gemeentenaam to keep the first entry
agg_dict_22['TOTAL'] = 'first'

# Perform the group by and aggregate functions
pivoted_MRA_sector_employment_pc4_2022 = pivoted_MRA_sector_employment_2022.groupby('pc4').agg(agg_dict_22).reset_index()

pivoted_MRA_sector_employment_pc4_2018


sbi_sectie_omschrijving   pc4  \
0                        1000   
1                        1011   
2                        1012   
3                        1013   
4                        1014   
..                        ...   
359                      8241   
360                      8242   
361                      8243   
362                      8244   
363                      8245   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                        0                            
1                                                                     1870                            
2                                                                     2330                            
3                                                                     5070                            
4                                                                     2180                            
..                                                                     ...                            
359                                                                     80                            
360                                                                    160                            
361                                                                    290                            
362                                                                    120                            
363                                                                    130                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                     0                            0   
1                                    80                         1880   
2                                    60                         1420   
3                                   600                         1250   
4                                   770                          230   
..                                  ...                          ...   
359                                  10                           20   
360                                  70                          100   
361                                 490                           60   
362                                  50                           40   
363                                  60                           60   

sbi_sectie_omschrijving  Detailhandel  \
0                                   0   
1                                 930   
2                                7820   
3                                1550   
4                                 460   
..                                ...   
359                                10   
360                              1120   
361                               560   
362                                40   
363                               120   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                                 0   
1                                                                 0   
2                                                                 0   
3                                                                 0   
4                                                                 0   
..                                                              ...   
359                                                               0   
360                                                               0   
361                                                               0   
362                                                               0   
363                                                               0   

sbi_sectie_omschrijving  Financiële instellingen  \
0                                              0   
1                                            110   

In [133]:
pivoted_MRA_sector_employment_pc4_2022

sbi_sectie_omschrijving   pc4  \
0                        1000   
1                        1011   
2                        1012   
3                        1013   
4                        1014   
..                        ...   
366                      8241   
367                      8242   
368                      8243   
369                      8244   
370                      8245   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                        0                            
1                                                                     2030                            
2                                                                     2290                            
3                                                                     6060                            
4                                                                     2750                            
..                                                                     ...                            
366                                                                     90                            
367                                                                    210                            
368                                                                    340                            
369                                                                    130                            
370                                                                    180                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                     0                            0   
1                                    90                         1830   
2                                    70                         1530   
3                                   560                         1320   
4                                   420                          330   
..                                  ...                          ...   
366                                  10                            0   
367                                 110                           80   
368                                 460                           60   
369                                  80                           30   
370                                 140                          120   

sbi_sectie_omschrijving  Detailhandel  \
0                                   0   
1                                 940   
2                                6770   
3                                1620   
4                                 500   
..                                ...   
366                                20   
367                              1010   
368                               700   
369                                60   
370                               210   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                                 0   
1                                                                 0   
2                                                                 0   
3                                                                 0   
4                                                                 0   
..                                                              ...   
366                                                               0   
367                                                               0   
368                                                               0   
369                                                               0   
370                                                               0   

sbi_sectie_omschrijving  Financiële instellingen  \
0                                              0   
1                                            140   

In [134]:
# There are 371 PC4s in 2022 and 364 in 2018. 
# Therefore we need to exclude these extra pc4s in 2022 as done in total jobs 

sector_common_pc4 = pivoted_MRA_sector_employment_pc4_2022['pc4'].isin(pivoted_MRA_sector_employment_pc4_2018['pc4'])
pivoted_MRA_sector_employment_pc4_2022 = pivoted_MRA_sector_employment_pc4_2022[sector_common_pc4]

pivoted_MRA_sector_employment_pc4_2022


sbi_sectie_omschrijving   pc4  \
0                        1000   
1                        1011   
2                        1012   
3                        1013   
4                        1014   
..                        ...   
366                      8241   
367                      8242   
368                      8243   
369                      8244   
370                      8245   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                        0                            
1                                                                     2030                            
2                                                                     2290                            
3                                                                     6060                            
4                                                                     2750                            
..                                                                     ...                            
366                                                                     90                            
367                                                                    210                            
368                                                                    340                            
369                                                                    130                            
370                                                                    180                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                     0                            0   
1                                    90                         1830   
2                                    70                         1530   
3                                   560                         1320   
4                                   420                          330   
..                                  ...                          ...   
366                                  10                            0   
367                                 110                           80   
368                                 460                           60   
369                                  80                           30   
370                                 140                          120   

sbi_sectie_omschrijving  Detailhandel  \
0                                   0   
1                                 940   
2                                6770   
3                                1620   
4                                 500   
..                                ...   
366                                20   
367                              1010   
368                               700   
369                                60   
370                               210   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                                 0   
1                                                                 0   
2                                                                 0   
3                                                                 0   
4                                                                 0   
..                                                              ...   
366                                                               0   
367                                                               0   
368                                                               0   
369                                                               0   
370                                                               0   

sbi_sectie_omschrijving  Financiële instellingen  \
0                                              0   
1                                            140   

In [135]:
# Ensuring consistency in datatypes and handling NaN values as done similarly for total employment data

#2018

# Step 1: Convert column to object type

pivoted_MRA_sector_employment_pc4_2018['pc4'] = pivoted_MRA_sector_employment_pc4_2018['pc4'].astype('int64')
pivoted_MRA_sector_employment_pc4_2018['gemeentecode'] = pivoted_MRA_sector_employment_pc4_2018['gemeentecode'].astype('int64')

# Step 2: Remove NaN values from 'pc4' and 'gemeentecode' column
pivoted_MRA_sector_employment_pc4_2018.dropna(subset=['pc4'])
pivoted_MRA_sector_employment_pc4_2018.dropna(subset=['gemeentecode'])



# Step 3: Remove rows where 'pc4' and 'gemeentecode' is 0
pivoted_MRA_sector_employment_pc4_2018 = pivoted_MRA_sector_employment_pc4_2018[pivoted_MRA_sector_employment_pc4_2018['pc4'] != 0]
pivoted_MRA_sector_employment_pc4_2018 = pivoted_MRA_sector_employment_pc4_2018[pivoted_MRA_sector_employment_pc4_2018['gemeentecode'] != 0]

# Verify the changes
print(pivoted_MRA_sector_employment_pc4_2018['pc4'].dtype)
print(pivoted_MRA_sector_employment_pc4_2018['pc4'].isnull().sum())
print((pivoted_MRA_sector_employment_pc4_2018['pc4']== 0).sum())

print('Gemeente Code')
print(pivoted_MRA_sector_employment_pc4_2018['gemeentecode'].dtype)
print(pivoted_MRA_sector_employment_pc4_2018['gemeentecode'].isnull().sum())
print((pivoted_MRA_sector_employment_pc4_2018['gemeentecode']== 0).sum())




#2022
# Step 1: Convert column to object type

pivoted_MRA_sector_employment_pc4_2022['pc4'] = pivoted_MRA_sector_employment_pc4_2022['pc4'].astype('int64')
pivoted_MRA_sector_employment_pc4_2022['gemeentecode'] = pivoted_MRA_sector_employment_pc4_2022['gemeentecode'].astype('int64')

# Step 2: Remove NaN values from 'pc4' and 'gemeentecode' column
pivoted_MRA_sector_employment_pc4_2022.dropna(subset=['pc4'])
pivoted_MRA_sector_employment_pc4_2022.dropna(subset=['gemeentecode'])



# Step 3: Remove rows where 'pc4' and 'gemeentecode' is 0
pivoted_MRA_sector_employment_pc4_2022 = pivoted_MRA_sector_employment_pc4_2022[pivoted_MRA_sector_employment_pc4_2022['pc4'] != 0]
pivoted_MRA_sector_employment_pc4_2022 = pivoted_MRA_sector_employment_pc4_2022[pivoted_MRA_sector_employment_pc4_2022['gemeentecode'] != 0]

# Verify the changes
print(pivoted_MRA_sector_employment_pc4_2022['pc4'].dtype)
print(pivoted_MRA_sector_employment_pc4_2022['pc4'].isnull().sum())
print((pivoted_MRA_sector_employment_pc4_2022['pc4']== 0).sum())

print('Gemeente Code')
print(pivoted_MRA_sector_employment_pc4_2022['gemeentecode'].dtype)
print(pivoted_MRA_sector_employment_pc4_2022['gemeentecode'].isnull().sum())
print((pivoted_MRA_sector_employment_pc4_2022['gemeentecode']== 0).sum())

int64
0
0
Gemeente Code
int64
0
0
int64
0
0
Gemeente Code
int64
0
0


C:\Users\rohan\AppData\Local\Temp\ipykernel_37356\2258832397.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pivoted_MRA_sector_employment_pc4_2022['pc4'] = pivoted_MRA_sector_employment_pc4_2022['pc4'].astype('int64')
C:\Users\rohan\AppData\Local\Temp\ipykernel_37356\2258832397.py:37: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pivoted_MRA_sector_employment_pc4_2022['gemeentecode'] = pivoted_MRA_sector_employment_pc4_2022['gemeentecode'].astype('int64')


In [136]:
# Renaming columns for clarity

pivoted_MRA_sector_employment_pc4_2018 =  pivoted_MRA_sector_employment_pc4_2018.rename(columns={'pc4': 'PC4 Code', 'TOTAL': 'Total Across Categories', 'gemeentecode' : 'Gemeente Code', 'gemeentenaam': 'Gemeente Name' })
pivoted_MRA_sector_employment_pc4_2022 =  pivoted_MRA_sector_employment_pc4_2022.rename(columns={'pc4': 'PC4 Code', 'TOTAL': 'Total Across Categories', 'gemeentecode' : 'Gemeente Code', 'gemeentenaam': 'Gemeente Name' })

pivoted_MRA_sector_employment_pc4_2018

sbi_sectie_omschrijving  PC4 Code  \
0                            1000   
1                            1011   
2                            1012   
3                            1013   
4                            1014   
..                            ...   
359                          8241   
360                          8242   
361                          8243   
362                          8244   
363                          8245   

sbi_sectie_omschrijving  Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                                        0                            
1                                                                     1870                            
2                                                                     2330                            
3                                                                     5070                            
4                                                                     2180                            
..                                                                     ...                            
359                                                                     80                            
360                                                                    160                            
361                                                                    290                            
362                                                                    120                            
363                                                                    130                            

sbi_sectie_omschrijving  Bouwnijverheid  Cultuur, sport en recreatie  \
0                                     0                            0   
1                                    80                         1880   
2                                    60                         1420   
3                                   600                         1250   
4                                   770                          230   
..                                  ...                          ...   
359                                  10                           20   
360                                  70                          100   
361                                 490                           60   
362                                  50                           40   
363                                  60                           60   

sbi_sectie_omschrijving  Detailhandel  \
0                                   0   
1                                 930   
2                                7820   
3                                1550   
4                                 460   
..                                ...   
359                                10   
360                              1120   
361                               560   
362                                40   
363                               120   

sbi_sectie_omschrijving  Extraterritoriale organisaties en lichamen  \
0                                                                 0   
1                                                                 0   
2                                                                 0   
3                                                                 0   
4                                                                 0   
..                                                              ...   
359                                                               0   
360                                                               0   
361                                                               0   
362                                                               0   
363                                                               0   

sbi_sectie_omschrijving  Financiële instellingen  \
0                                              0   
1  

### Data Processing for both total and sectorwise job data is done at this point ###

# Creating Master Data at PC4 level #

In [137]:
# First Ensuring that the total pc4 zones in the LISA data provided by O&S team at Amsterdam match with the CBS PC4 data


#Need to use CBS data for 2023, Because it is updated till august 2023. same reason as for considering jobs 
# We used the 2023 data for jobs for the same reason because its till April 2023 and ODIN data can be any time in the year.

In [138]:
#First replace GM with nothing in the excel file and save it!

pc6_CBS_file_2023 = '2b-GWB2023_PC6.xlsx'
pc6_CBS_2023 = pd.read_excel(pc6_CBS_file_2023)
pc6_CBS_2023

GemCode         GemNaam  WijkCode                      Wijknaam  \
0            14       Groningen  WK001400                       Centrum   
1            14       Groningen  WK001400                       Centrum   
2            14       Groningen  WK001400                       Centrum   
3            14       Groningen  WK001400                       Centrum   
4            14       Groningen  WK001400                       Centrum   
...         ...             ...       ...                           ...   
493159     1992  Voorne aan Zee  WK199232                    Zwartewaal   
493160     1992  Voorne aan Zee  WK199233  Recreatiestrook Brielse Maas   
493161     1992  Voorne aan Zee  WK199233  Recreatiestrook Brielse Maas   
493162     1992  Voorne aan Zee  WK199233  Recreatiestrook Brielse Maas   
493163     1992  Voorne aan Zee  WK199233  Recreatiestrook Brielse Maas   

         BuurtCode                     Buurtnaam     PC6  AantalAdressen  
0       BU00140000              Binnenstad-Noord  9711LX               1  
1       BU00140000              Binnenstad-Noord  9712AA               9  
2       BU00140000              Binnenstad-Noord  9712AB               6  
3       BU00140000              Binnenstad-Noord  9712AC              12  
4       BU00140000              Binnenstad-Noord  9712AD              15  
...            ...                           ...     ...             ...  
493159  BU19923209             Verspreide huizen  3238LZ             163  
493160  BU19923300  Recreatiestrook Brielse Maas  3231NA               3  
493161  BU19923300  Recreatiestrook Brielse Maas  3231NC               3  
493162  BU19923300  Recreatiestrook Brielse Maas  3231ND              12  
493163  BU19923300  Recreatiestrook Brielse Maas  3231NE               5  

[493164 rows x 8 columns]

In [139]:
pc6_CBS_2023.columns

Index(['GemCode', 'GemNaam', 'WijkCode', 'Wijknaam', 'BuurtCode', 'Buurtnaam',
       'PC6', 'AantalAdressen'],
      dtype='object')

In [140]:
pc6_CBS_2023 =  pc6_CBS_2023[['PC6', 'GemCode', 'GemNaam']]

# Renaming retained column names for clarity, PC6 is renamed to PC4 code as it will be converted to PC4 later
pc4_CBS_2023 = pc6_CBS_2023.rename(columns = {'PC6': 'PC4 Code', 'GemCode' : 'Gemeente Code',  'GemNaam' : 'Gemeente Name' })
pc4_CBS_2023['PC4 Code'] = pc4_CBS_2023['PC4 Code'].astype(str).str[:4] # Retaining only the first 4 digit of PC4( Converting it to PC4)
pc4_CBS_2023['PC4 Code'] =  pc4_CBS_2023['PC4 Code'].astype('int64') #Converting it to int type for consistency with the processed LISA data

pc4_CBS_2023


PC4 Code  Gemeente Code   Gemeente Name
0           9711             14       Groningen
1           9712             14       Groningen
2           9712             14       Groningen
3           9712             14       Groningen
4           9712             14       Groningen
...          ...            ...             ...
493159      3238           1992  Voorne aan Zee
493160      3231           1992  Voorne aan Zee
493161      3231           1992  Voorne aan Zee
493162      3231           1992  Voorne aan Zee
493163      3231           1992  Voorne aan Zee

[493164 rows x 3 columns]

In [141]:

# First filtering the PC4 only in the MRA region
MRA_pc4_CBS_2023 =  pc4_CBS_2023[pc4_CBS_2023['Gemeente Code'].isin(MRA_total_employment_pc4_2022['Gemeente Code'])]


# Since mulitple PC6 will have the same PC4 (only the last 2 letters differ) retaining only the first one
MRA_pc4_CBS_2023 = MRA_pc4_CBS_2023.groupby('PC4 Code').agg({
    'Gemeente Code': 'first',
    'Gemeente Name': 'first'
}).reset_index()
MRA_pc4_CBS_2023

PC4 Code  Gemeente Code Gemeente Name
0        1011            363     Amsterdam
1        1012            363     Amsterdam
2        1013            363     Amsterdam
3        1014            363     Amsterdam
4        1015            363     Amsterdam
..        ...            ...           ...
359      8241            995      Lelystad
360      8242            995      Lelystad
361      8243            995      Lelystad
362      8244            995      Lelystad
363      8245            995      Lelystad

[364 rows x 3 columns]

In [142]:
#Checking which PC4s are not common across LISA provided by O&S and CBS data


MRA_CBS_pc4_23set = set(MRA_pc4_CBS_2023['PC4 Code'])
MRA_Lisa_pc4_23set = set(MRA_total_employment_pc4_2022['PC4 Code'])


pc4_not_in_Lisa = MRA_CBS_pc4_23set - MRA_Lisa_pc4_23set

print('pc4 not in Lisa:', pc4_not_in_Lisa)
print(len(pc4_not_in_Lisa))


pc4_not_in_CBS =  MRA_Lisa_pc4_23set - MRA_CBS_pc4_23set

print('pc4 not in CBS:',pc4_not_in_CBS)
print(len(pc4_not_in_CBS))


pc4 not in Lisa: {1364}
1
pc4 not in CBS: {1000}
1


In [143]:
# If no job info is avaialble in LISA for 1364, exclude that PC4 altogether, 
# and 1000 too has only 1 job, so can be excluded too
#MRA_total_employment_pc4_2018[MRA_total_employment_pc4_2018['PC4 Code'] == 1364]
#MRA_total_employment_pc4_2018[MRA_total_employment_pc4_2018['PC4 Code'] == 1000]


In [144]:
# Merging CBS and LISA datasets based on PC4s in the 2023 CBS data,


#2018 
Masterdata_MRA_2018 = pd.merge(
MRA_pc4_CBS_2023,
MRA_total_employment_pc4_2018 [['PC4 Code', 'Total Employment 2018-19']],

how = 'left',
left_on = 'PC4 Code',
right_on = 'PC4 Code',

)
# Replace NaN with 0 and convert to int64
Masterdata_MRA_2018['Total Employment 2018-19'] = Masterdata_MRA_2018['Total Employment 2018-19'].fillna(0).astype('int64')



#2022
Masterdata_MRA_2022 = pd.merge(
MRA_pc4_CBS_2023,
MRA_total_employment_pc4_2022 [['PC4 Code', 'Total Employment 2022-23']],

how = 'left',
left_on = 'PC4 Code',
right_on = 'PC4 Code',

)
# Replace NaN with 0 and convert to int64
Masterdata_MRA_2022['Total Employment 2022-23'] = Masterdata_MRA_2022['Total Employment 2022-23'].fillna(0).astype('int64')

In [145]:
pivoted_MRA_sector_employment_pc4_2018.columns

Index(['PC4 Code',
       'Advisering, onderzoek en overige specialistische zakelijke dienstverlening',
       'Bouwnijverheid', 'Cultuur, sport en recreatie', 'Detailhandel',
       'Extraterritoriale organisaties en lichamen', 'Financiële instellingen',
       'Gezondheids- en welzijnszorg', 'Groothandel', 'Industrie',
       'Informatie en communicatie', 'Landbouw, bosbouw en visserij',
       'Logies-, maaltijd- en drankverstrekking', 'Onderwijs',
       'Openbaar bestuur, overheidsdiensten en verplichte sociale verzekeringen',
       'Overige dienstverlening',
       'Productie en distributie van en handel in elektriciteit, aardgas, stoom en gekoelde lucht',
       'Verhuur van en handel in onroerend goed',
       'Verhuur van roerende goederen en overige zakelijke dienstverlening',
       'Vervoer en opslag',
       'Winning en distributie van water: afval- en afvalwaterbeheer en sanering',
       'Winning van delfstoffen ', 'Gemeente Code', 'Gemeente Name',
       'Total Across 

In [146]:
columns_to_sum_18

['Advisering, onderzoek en overige specialistische zakelijke dienstverlening',
 'Bouwnijverheid',
 'Cultuur, sport en recreatie',
 'Detailhandel',
 'Extraterritoriale organisaties en lichamen',
 'Financiële instellingen',
 'Gezondheids- en welzijnszorg',
 'Groothandel',
 'Industrie',
 'Informatie en communicatie',
 'Landbouw, bosbouw en visserij',
 'Logies-, maaltijd- en drankverstrekking',
 'Onderwijs',
 'Openbaar bestuur, overheidsdiensten en verplichte sociale verzekeringen',
 'Overige dienstverlening',
 'Productie en distributie van en handel in elektriciteit, aardgas, stoom en gekoelde lucht',
 'Verhuur van en handel in onroerend goed',
 'Verhuur van roerende goederen en overige zakelijke dienstverlening',
 'Vervoer en opslag',
 'Winning en distributie van water: afval- en afvalwaterbeheer en sanering',
 'Winning van delfstoffen ']

In [147]:
# Verifying if the data processing is done correctly, 
# by checking if jobs across various sectors match close to the total job data for a PC4



# Creating a 'Total Across Categories' by summing up jobs across all categories
Masterdata_MRA_2018 = pd.merge(
Masterdata_MRA_2018,
pivoted_MRA_sector_employment_pc4_2018[['PC4 Code', 'Total Across Categories']+columns_to_sum_18 ],

how = 'left',
on = 'PC4 Code',

)





# Replace NaN with 0 and convert to int64
Masterdata_MRA_2018[columns_to_sum_18] = Masterdata_MRA_2018[columns_to_sum_18].fillna(0).astype('int64')
Masterdata_MRA_2018['Total Across Categories'] = Masterdata_MRA_2018['Total Across Categories'].fillna(0).astype('int64')



In [148]:
pivoted_MRA_sector_employment_pc4_2022.columns

Index(['PC4 Code',
       'Advisering, onderzoek en overige specialistische zakelijke dienstverlening',
       'Bouwnijverheid', 'Cultuur, sport en recreatie', 'Detailhandel',
       'Extraterritoriale organisaties en lichamen', 'Financiële instellingen',
       'Gezondheids- en welzijnszorg', 'Groothandel', 'Industrie',
       'Informatie en communicatie', 'Landbouw, bosbouw en visserij',
       'Logies-, maaltijd- en drankverstrekking', 'Onderwijs',
       'Openbaar bestuur, overheidsdiensten en verplichte sociale verzekeringen',
       'Overige dienstverlening',
       'Productie en distributie van en handel in elektriciteit, aardgas, stoom en gekoelde lucht',
       'Verhuur van en handel in onroerend goed',
       'Verhuur van roerende goederen en overige zakelijke dienstverlening',
       'Vervoer en opslag',
       'Winning en distributie van water: afval- en afvalwaterbeheer en sanering',
       'Winning van delfstoffen ', 'Gemeente Code', 'Gemeente Name',
       'Total Across 

In [149]:
#Same operation for 2022

Masterdata_MRA_2022 = pd.merge(
Masterdata_MRA_2022,
pivoted_MRA_sector_employment_pc4_2022[['PC4 Code', 'Total Across Categories']+columns_to_sum_22 ],

how = 'left',
on = 'PC4 Code',

)

Masterdata_MRA_2022[columns_to_sum_22] = Masterdata_MRA_2022[columns_to_sum_22].fillna(0).astype('int64')
Masterdata_MRA_2022['Total Across Categories'] = Masterdata_MRA_2022['Total Across Categories'].fillna(0).astype('int64')

In [150]:
Masterdata_MRA_2022

PC4 Code  Gemeente Code Gemeente Name  Total Employment 2022-23  \
0        1011            363     Amsterdam                     14203   
1        1012            363     Amsterdam                     31692   
2        1013            363     Amsterdam                     23203   
3        1014            363     Amsterdam                     14759   
4        1015            363     Amsterdam                      9241   
..        ...            ...           ...                       ...   
359      8241            995      Lelystad                       235   
360      8242            995      Lelystad                      2452   
361      8243            995      Lelystad                      5432   
362      8244            995      Lelystad                       669   
363      8245            995      Lelystad                      1204   

     Total Across Categories  \
0                      14170   
1                      31680   
2                      23210   
3                      14760   
4                       9240   
..                       ...   
359                      230   
360                     2430   
361                     5450   
362                      660   
363                     1220   

     Advisering, onderzoek en overige specialistische zakelijke dienstverlening  \
0                                                 2030                            
1                                                 2290                            
2                                                 6060                            
3                                                 2750                            
4                                                 2260                            
..                                                 ...                            
359                                                 90                            
360                                                210                            
361                                                340                            
362                                                130                            
363                                                180                            

     Bouwnijverheid  Cultuur, sport en recreatie  Detailhandel  \
0                90                         1830           940   
1                70                         1530          6770   
2               560                         1320          1620   
3               420                          330           500   
4               180                          990           750   
..              ...                          ...           ...   
359              10                            0            20   
360             110                           80          1010   
361             460                           60           700   
362              80                           30            60   
363             140                          120           210   

     Extraterritoriale organisaties en lichamen  ...  \
0                                             0  ...   
1                                             0  ...   
2                                             0  ...   
3                                             0  ...   
4                                             0  ...   
..                                          ...  ...   
359                                           0  ...   
360                                           0  ...   
361                                           0  ...   
362                                           0  ...   
363                                           0  ...   

     Logies-, maaltijd- en drankverstrekking  Onderwijs  \
0                                       1560       1240   
1                                       9740       1560   
2                                       1410        700   
3                                        610        790 

In [151]:
Masterdata_MRA_2018.to_csv('Masterdata Employment MRA 2018.csv', index=False)

Masterdata_MRA_2022.to_csv('Masterdata Employment MRA 2022.csv', index=False)

# Master data including Total and Sector wise Jobs per PC4 done! #

# Next up: Education